### ISSUE TITLE
Rebuild is way slower and stop working after several rebuilds

### ISSUE THREAD

**Issue by montella1507** [angular/angular-cli#20235]
### Affected Package
Propably compiler 11.2.5

We have migrated from:
```
  "@angular/animations": "^11.0.9",
    "@angular/cdk": "^11.0.4",
    "@angular/common": "^11.0.9",
    "@angular/compiler": "^11.0.9",
    "@angular/core": "^11.0.9",
  "@angular/forms": "^11.0.9",
"@angular/platform-browser": "^11.0.9",
    "@angular/platform-browser-dynamic": "^11.0.9",
    "@angular/router": "^11.0.9",
    "@angular/service-worker": "^11.0.9"
  "@angular-devkit/build-angular": "^0.1100.7",
  "@angular/cli": "^11.0.7",
    "@angular/compiler-cli": "^11.0.9",
    "@angular/language-service": "^11.0.9"
```

TO 

```
 "@angular/animations": "^11.2.5",
    "@angular/cdk": "^11.2.3",
    "@angular/common": "^11.2.5",
    "@angular/compiler": "^11.2.5",
    "@angular/core": "^11.2.5",
  "@angular/forms": "^11.2.5",
 "@angular/platform-browser": "^11.2.5",
    "@angular/platform-browser-dynamic": "^11.2.5",
    "@angular/router": "^11.2.5",
    "@angular/service-worker": "^11.2.5",
  "@angular-devkit/build-angular": "^0.1102.3",
"@angular/cli": "^11.2.3",
    "@angular/compiler-cli": "^11.2.5",
    "@angular/language-service": "^11.2.5"

```
### Is this a regression?
YES

### Description
We have migrated from angular 11.0.9 to 11.2.5. Our application is pretty big 1300components, 170 NgModules,  70 lazy loaded chunks. 

After migrations almost all rebuild times took 6-10x more time and what is worse.. 
1) Rebuild starts afer XYZms (sometimes several seconds) after save of the file
2) Rebuild stopped working after 2-3 saves (rebuilds) , even when change is a breaking change (return 0 to return 1 in component method).

I am willling to help with debug (profile node process / generate and send any log, however it may be imposible to prepare minimal repro :-/ )

**Comment by montella1507** [comment]
Cool :-) I will try with AOT and will come back with report :-)

**Comment by montella1507** [comment]
[vscode-profile-2021-03-10-21-36-15-deep-component-change.zip](https://github.com/angular/angular-cli/files/6118723/vscode-profile-2021-03-10-21-36-15-deep-component-change.zip)

Rebuild propably will not stop working, however rebuild times still super slow :-(

**Comment by JoostK** [comment]
Ok, thanks for the additional profile. I do indeed see a lot more emit activity than I'd suspect, but I can't tell why that would be the case.

Another observation is that the cyclic import check takes 17s, so you may want to disable that in devbuilds (there's a setting in `angular.json`, not sure what it's called from the top of my head).

**Comment by JoostK** [comment]
@montella1507 Could you also try with `NG_BUILD_IVY_LEGACY=1`? I doubt that it will help us diagnose the cause of the issue, but hopefully it does allow you to get back reasonable incremental rebuilds.

**Comment by JoostK** [comment]
Thanks for the additional information. I suspect that it's just Webpack's dependency graph that makes it believe that everything needs to be reemitted, which is introducing a huge bottleneck as emit is the slowest of all phases. The improvements to 11.2.5 are only concerning Angular's involvement in the process, which used to be suboptimal, but given that your JIT recompiles were also slow means that this is not the relevant here (my initial comment about JIT mode being affected by the AOT compiler turned out to be inaccurate). So that leaves a new problem to be investigated with respect to the dependency graph that Webpack is using to drive recompiles.

**Comment by montella1507** [comment]
```
set NG_BUILD_IVY_LEGACY=1
ng serve...
```

mode: JIT
cyclic: DISABLED

rebuild stiill like 5-6 times slower than in previous NG version (but it works atleast even after 10th change/rebuild).  
Interesting thing here is, that first  few (1-2) rebuilds it changes 4+ chunks, but 3rd+ rebuild correctly rebuilds only one. (so rebuild time drops from 30sec to 7sec).  

However i am not able to profile (profiller / debugging and entire build will crash)

**Comment by montella1507** [comment]
[vscode-profile-2021-03-10-23-09-20.zip](https://github.com/angular/angular-cli/files/6119127/vscode-profile-2021-03-10-23-09-20.zip)

Ah finaly:
 NG_BUILD_IVY_LEGACY=1
 MODE: JIT
 cyclic: disabled


So most time takes source map generation i think.

**Comment by montella1507** [comment]
@JoostK  hi, btw i have made additional benchmarks on 11.0.9 to measure circular-dependency-plugin  + sourcmap plugin effects on build time.. here are the results:

![image](https://user-images.githubusercontent.com/10418323/110759583-0f425f00-824e-11eb-820a-eadf082273b1.png)


When circular-dependency detection is enabled, build and rebuild takes 5-8x more time. I think it have serious issues with async circular dependencies or something... 

And yes,       "showCircularDependencies": false is enough. Will do the same benchmarks on current version of angular / compiler later (when this issue is closed or requested by anyone explicitely)

**Comment by montella1507** [comment]
![image](https://user-images.githubusercontent.com/10418323/110761739-5f222580-8250-11eb-8c70-e80891a10af5.png)
"bez cyklic = no circular-dependency-plugin"

**Comment by JoostK** [comment]
@montella1507 We think to have found the culprit, it was fixed on master in angular/angular-cli#20236 and will be ported to patch so should be part of next week's release. I'll share a snapshot build for patch once that's available.

**Comment by montella1507** [comment]
Thank you very much.  :-)  

I have realized rebuild is affected by assets too - created this issue - maybe it is duplicit to this one - feel free to close please if neccessary.

https://github.com/angular/angular-cli/issues/20244

**Comment by JoostK** [comment]
Snapshot build is here: https://github.com/angular/ngtools-webpack-builds/tree/11.2.x (see top of readme)

**Comment by clydin** [comment]
That snapshot build includes fixes for the repeat JIT build crash issue (fix via #20237) and the rebuild performance (fix via #20247).  If you have the opportunity, please give the snapshot build a try and let us know if you are still encountering issues.

Also, the asset issue is due to a third-party dependency used by the Angular CLI so a separate issue is appropriate.  Some more details regarding the asset performance in serve can be found in a comment in that issue.

**Comment by montella1507** [comment]
Thank you will try soon (1-2 days - propably weekend). BTW are you willing to consider to disable circular-dependency-plugin as default? TBH it is cause of serious DX  issues on medium - large projects. I tried to disable that on medium projects without dynamic cross references (in comments above) and it affects rebuild time too (30% in medium project, 80% on large with dynamic cross-references). 

Or would you accept my PR about "rebuild-time performance" article in documentation?

**Comment by clydin** [comment]
The `circular-dependency-plugin` default is definitely something the team will consider.  It is, however, a breaking change to modify the default so timing of the change will need to be carefully considered.

**Comment by montella1507** [comment]
```
npm install git+https://github.com/angular/ngtools-webpack-builds.git
+ @ngtools/webpack@12.0.0-next.4
added 6 packages from 9 contributors, removed 1 package, updated 1 package and audited 3255 packages in 27.004s
(however crashed nody-gyp build - but i think that doesnt matter)
```


**JIT:**
First build:
Build at: 2021-03-11T15:06:08.464Z - Hash: bf982f060453b54ec149 - Time: 262241ms
Circular-dependency-plugin: disabled
(JIT without that LEGACY IVY flag)

App.component.ts   (it took 3.5sec on  11.0.9)
rebuild       time
1                 36 192ms
2                 29 191 ms
3                 24 857 ms 

**AOT:**
First build:
Build at: 2021-03-11T15:06:08.464Z - Hash: bf982f060453b54ec149 - Time: 262241ms
Circular-dependency-plugin: disabled
(AOT without that LEGACY IVY flag)

App.component.ts   (it took 3.5sec on  11.0.9)
rebuild       time
1                 36 192ms
2                 29 191 ms
3                 24 857 ms 

So i am not sure it had any effect (can i check somehow i am actually using that snapshot build? i have that "function augmentHostWithDependencyCollection" in my node_modules so i expect it is correctly updated to snapshot build) Here is the CPU profile file (AOT, AppComponent rebuild)  





First build:
Build at: 2021-03-11T15:13:27.066Z - Hash: bdf37086378ab2e2bfca - Time: 179718ms
Circular-dependency-plugin: disabled
(AOT without that LEGACY IVY flag)

App.component.ts   
rebuild       time
1                 50 400ms
2                 44 000ms
3                 47 600ms


[snapshot-profiling.zip](https://github.com/angular/angular-cli/files/6123998/snapshot-profiling.zip)
(took so much time on pipelineEmitWithCommends and _readdirp)

**Comment by clydin** [comment]
Unfortunately, the `@ngtools/webpack` package is a direct dependency of `@angular-devkit/build-angular` so it won't be used if installed directly in the project.  If using yarn, a `resolutions` addition in the project's `package.json` would force the version used.  For npm, you could try manually removing any instances of `@ngtools/webpack` from `node_modules` that are not the snapshot version (there should be one inside the node modules directory of `@angular-devkit/build-angular`).

**Comment by montella1507** [comment]
OK renamed @ngtools/webpack in @angular-devkit/build-angular folder, ls output:


```
+-- @angular-devkit/build-angular@0.1102.3
| `-- UNMET DEPENDENCY @ngtools/webpack@11.2.3
+-- @ngtools/webpack@12.0.0-next.4  (git+https://github.com/angular/ngtools-webpack-builds.git#dafcd5969c31e41f7a1893ebb83cc8ebe53cf6da)
`-- @nrwl/builders@7.8.7
  `-- @angular-devkit/build-angular@0.13.10
    `-- @ngtools/webpack@7.3.10
```

**Comment by clydin** [comment]
Just noticed that you are in a v11 project but the plugin is for a v12 prelease.  For a v11 version of the snapshot with the commits you can use:
```
npm install git+https://github.com/angular/ngtools-webpack-builds.git#ba719cff4bab883b91684738b04bea0145ec2a81
```

**Comment by montella1507** [comment]
**Results - AOT - no circular**:
App.component.ts
rebuild time
1st        11 994ms
2nd      10 900ms
3rd         9  000ms 

Thats very impressive  🤟 
**Results - JIT - no circular:**

App.component.ts 
rebuild time
BEST  5 500ms now / 3 600ms on 11.0.9

Col.component.ts (deep shared component)
BEST 6 200 ms now /  3700md on 11.0.9

User-role.component.ts (page component)
BEST 4300 ms now /  1950ms on 11.0.9

Stock.component.ts 
BEST 4000ms now / 2100ms on 11.0.9

So i can confirm 3 things:
- it is like 6-8x faster than stock 11.2.5
- +- 2x slower than 11.0.9
- it works (no broken rebuild)

Here is the profile from JIT 11.2.5-next.4

[vscode-profile-2021-03-11-16-41-25.zip](https://github.com/angular/angular-cli/files/6124149/vscode-profile-2021-03-11-16-41-25.zip)

**Comment by JoostK** [comment]
Can you share an AOT profile as well? Thanks for all the insights btw, super helpful :+1:

**Comment by montella1507** [comment]
TBH if you want to, i can provide you remote desktop to my computer and you can do anything you want to 😄  (keeping that as LAST chance how to realize the problem ). Unfortunately cannot share the project files bcs of NDA.

Here is the AOT profile:

[next-5-aot-col-component.zip](https://github.com/angular/angular-cli/files/6124223/next-5-aot-col-component.zip)

**Comment by JoostK** [comment]
Great, thanks. What version of Node are you using?

**Comment by montella1507** [comment]
v14.15.1
npm 6.14.8

**Comment by montella1507** [comment]
Projects are on NVMe Samsung PRO SSDs
Intel(R) Core(TM) i7-10700K CPU @ 3.80GHz   3.79 GHz ( 4.7 boost stable)
32GB 3600Mhz Dual channel RAMs

**Comment by JoostK** [comment]
Cool. It doesn't get much faster than that (except for Windows, which is generally slow for I/O). There's quite some interesting things in the profile, one of which can probably be improved on our side.

**Comment by dgp1130** [comment]
We discussed this issue in our weekly triage meeting. We also did a bit of digging and found the original PR (#6813) and it seems the main motivation was to fix a compiler bug (#6309). That bug only occurred with the "magic import string" which is no longer supported. The compiler has also evolved a lot since then, so this is unlikely to still be an issue. The circular dependency check is also only a warning, so I don't think it really fixed the compiler problem, just gave a helpful error message about the problem.

General consensus is that the circular dependency checker doesn't provide a whole lot of value. While circular dependencies can cause issues, most of the Node toolchain supports them so it isn't a major problem for most web projects. Angular also supports circular dependencies<sup>[[citation needed]](https://xkcd.com/285/)</sup>. The compiler throws errors in specific, problematic cases but generally allows circular dependencies. This is only an issue if a project owner decides it is an issue. As a result, it is reasonable for them to integrate such a check without necessarily expecting one out of the box. [ESLint has one such check](https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-cycle.md), and there are likely other implementations as well.

We could just run this check for production builds, however prod builds are usually done by CI and the circular dependency check is only a warning, so it is unlikely many humans would see this log and take any meaningful action from it. Since the check provides little value now, limiting it to production builds will only reduce that value further. As a result, we're looking into removing the circular dependency check altogether. This will speed up build times (as shown above), reduce third party dependencies, and and `node_modules/` a bit smaller.

Two quick things to confirm before we commit to removing the plugin:

1. @filipesilva, you originally added the circular dependency check in #6813 (thanks for the detailed commit message!), is there any other context or considerations we should have before removing it? Any objections to these arguments?
2. We should also include a proper circular dependency test case with our new dynamic import infrastructure if it does not already exist. This would also confirm that the previous compiler issue no longer exists and that it is safe to remove the circular dependency check.

**Comment by montella1507** [comment]
`We could just run this check for production builds, however prod builds are usually done by CI and the circular dependency check is only a warning, so it is unlikely many humans would see this log and take any meaningful action from it. `

we have modified CI configuration that it runs build from commits **WITHOUT** Source-maps and **WITHOUT** circular dependency check  +  AUTO night builds (any periodic builds) runs WITH SM and WITH CDCH.  

It changed time required to 1/10 and still viable option to detect circular dependencies (it is very rare situation for us - we have good barel files strategies + NRWL tags + linters).

**Comment by montella1507** [comment]
@JoostK  @dgp1130  hi guys, should i close this issue?

**Comment by clydin** [comment]
@montella1507 If you have some additional time, could you try the latest snapshot build?
```
npm install git+https://github.com/angular/ngtools-webpack-builds.git#a2c144fbeeef9c9584b7848acdb0d1f5b4d9e20b
```
You'll need to do the same manual change to node modules again to force the `@ngtools/webpack` usage.

Angular CLI 11.2.5 will be released this week but feedback before the release would be helpful.
Thank you again for all the profiling information.

**Comment by JoostK** [comment]
@montella1507 I'd be okay to keep it open until after the next release (likely today), so that you can check whether it's back to normal then.

**Comment by montella1507** [comment]
@clydin  npm ls @ngtools/webpack output
`
+-- @angular-devkit/build-angular@0.1100.7
| `-- UNMET DEPENDENCY @ngtools/webpack@11.0.7
+-- @ngtools/webpack@11.2.4  (git+https://github.com/angular/ngtools-webpack-builds.git#a2c144fbeeef9c9584b7848acdb0d1f5b4d9e20b)
`-- @nrwl/builders@7.8.7
  `-- @angular-devkit/build-angular@0.13.10
    `-- @ngtools/webpack@7.3.10
`
### AOT: FALSE / CIRCULAR PLUGIN: TRUE (best rebuild times)
App.component (main chunk) - 12 156 
COL.component (deep shared - main chunk)  - 10 400
user-role  8 500
stock-component (routed components - lazy chunk) 9400 

### AOT: FALSE / CIRCULAR PLUGIN: FALSE (best rebuild times)
App.component (main chunk) - 3 500
COL.component (deep shared - main chunk)  - 3300
user-role  2 000
stock-component (routed components - lazy chunk) 3000 

### AOT: TRUE / CIRCULAR PLUGIN: FALSE (best rebuild times)
App.component (main chunk) - 12 000
COL.component (deep shared - main chunk)  -     48 000 ms (not cool :-/) 
user-role  11 000
stock-component (routed components - lazy chunk) 11 000

WILL check times again when you publish new version (with clean install of packages - just to be sure everything is UP TO DATE)
 
Here is the CPU PROFILE FROM - AOT: TRUE / CIRCULAR PLUGIN: FALSE / COL.component (48 000ms rebuild). There MAY be prblem with packages (i am not sure everything is up to date), but you would see in cpu profile:
[vscode-profile-2021-03-17-20-19-28.zip](https://github.com/angular/angular-cli/files/6159161/vscode-profile-2021-03-17-20-19-28.zip)

**Comment by clydin** [comment]
From the profile it looks like the legacy Ivy plugin is being used along with an older version of the `@angular/compiler-cli`.  That would probably explain those varying timings.

The next patch release of the Angular CLI should be out today.  So it may be better to wait and try a full update to the latest versions of 11.2.x.

**Comment by JoostK** [comment]
Another observation is that you're not yet on TypeScript 4.1. This affects source-map performance in AOT compilations, as there was a severe performance penalty for source mapping into external templates.

**Comment by montella1507** [comment]
OK. Just let me know when everything is OUT. I will:
1) pull latest source code
2) ng update
3) rm package-lock
4) npm i
5) provide correct results with everything up to date 
:-)

**Comment by JoostK** [comment]
@montella1507 Refresh once more 😄

So just to summarize: you'll need `@angular/compiler-cli` (and its other framework friends) at version >= 11.2.5, with 11.2.6 being today's release. Additionally, you'll need `@angular/cli` 11.2.5, i.e. today's release. Furthermore, it would be beneficial to use TS 4.1 if combined with AOT, but I don't expect this to matter much in incremental rebuilds (only initial builds).

**Comment by dgp1130** [comment]
@montella1507 CLI `11.2.5` just released, so you should be unblocked there.

**Comment by montella1507** [comment]
I have problem with updating to TS 4.1, because:
- i found this PR "update to TS 4.1" - in commits is change to 4.1.2
- 4.1.0 doesnt exist, do you mean  4.1.2, 4.1.5 or  4.2.3 ?
- angular has   "typescript": "~4.2.3", as dependency
- angular/cli has     "typescript": "4.2.3", as dependency...

so what should i change to? 4.2.3 ?

**Comment by JoostK** [comment]
Don't worry too much about the TS update. I meant TS 4.1.x in general, so that would be the latest 4.1.5. You can use `~4.1.5` as version specifier to get the latest 4.1.x but not 4.2.x (TS does not follow semver). The ~4.2.3 dependency version is for the prereleases of Angular 12.

**Comment by montella1507** [comment]
What am i doing wrong? :-O

npm install typescript 4.1.5 --skip-cache
npm ERR! code E404
npm ERR! 404 Not Found - GET https://registry.npmjs.org/4.1.5 - Not found

**Comment by montella1507** [comment]
ok thanks :-) npm install typescript@4.1.5 --skip-cache  

will paste results soon.

**Comment by JoostK** [comment]
It could be that the TS update introduces some new compilation errors, so you can choose to leave that one for now and just update to latest Angular. The effect of the TS improvements are mostly noticeable on initial builds.

**Comment by montella1507** [comment]
1) maybe it is bug that i see 0 bytes in rebuilt chunk size ? :-)
```
Lazy Chunk Files                                        | Names                                                |    Size      
components-src-lib-user-management-components-module.js | components-src-lib-user-management-components-module | 0 bytes  
```

2) Let me know when you got to Czech Republic, i owe you a beer

![image](https://user-images.githubusercontent.com/10418323/111554323-65e7f580-8786-11eb-96b9-f34d065192d6.png)
![image](https://user-images.githubusercontent.com/10418323/111554332-6a141300-8786-11eb-9543-e6d3710bf4b9.png)


NOTES:
- in 11.0.9, disabling Sourcemaps means not big improvement in JIT mode, in 11.2.6 it means often 50%
- 11.2.6 improved ALL times speciallz AOT rebuild time in deep-shared component

**Comment by JoostK** [comment]
That's more like it :-)

The AOT improvement to deep-shared component is due to angular/angular#40947.

Which version of TS was this? And could you share an AOT / NO CIRC / NO SM profile as well?

**Comment by montella1507** [comment]
Fortunately, i have jus got enabled 4H of Zack Snyder's Justice League CUT so i have enough time to profile on second display :-)   TS. is 4.1.5 Profiles will provide soon.

**Comment by montella1507** [comment]
[nosm-nocirc-aot-multiple-components.zip](https://github.com/angular/angular-cli/files/6160525/nosm-nocirc-aot-multiple-components.zip)
here we go.

**Comment by JoostK** [comment]
Thanks, those are all JIT though. A single variant would do, they should be similar enough.

**Comment by montella1507** [comment]
[aot.zip](https://github.com/angular/angular-cli/files/6160591/aot.zip)

**Comment by jlabaj** [comment]
even after the update to 15.2.5 it still takes 40s instead of 5 for example without the env. variable: NG_BUILD_IVY_LEGACY=1 

No maps no aot. I also updated the typescript to 4.1

**Comment by clydin** [comment]
Please provide the output of “ng version”.

**Comment by jlabaj** [comment]
@clydin 
![image](https://user-images.githubusercontent.com/5179937/111611840-2bf60e00-87dd-11eb-9e24-2f720a55d964.png)

**Comment by clydin** [comment]
There appears to be mixture of package versions in the project; some of which are incompatible with each other and some are deprecated.
Please open a new issue with the above information as well as the contents of the “package.json” file and whether “ng update” was used to update versions (major or otherwise).

**Comment by jlabaj** [comment]
@clydin here you go #20319

**Comment by montella1507** [comment]
> Thanks, those are all JIT though. A single variant would do, they should be similar enough.

I have uploaded AOT.zip with aot profiles, let me know if i can help you with anything else.  If you think it is already solved - and you okay with results, i can close the issue (from my POV)

**Comment by jlabaj** [comment]
@clydin reinstalling the @angular-builders/custom-webpack@11.0.0 helped, updating to @angular-builders/custom-webpack@11.1.1 continued to work.

The build time with sourcemaps still did not improve however, still 40s. also mentioned here #20319

**Comment by JoostK** [comment]
@montella1507 Thanks for the profile. It's mostly what I was expecting from the earlier profiles and the changes that have been made since.

If you feel like it, can you run another experiment for me? Based on your profiling output I have been experimenting with cycle analysis (not the expensive step that you can disable, but an internal Angular compiler analysis step needed for proper output) and I'd like to know what the impact of that would be.

The [build artifacts are here](https://app.circleci.com/pipelines/github/angular/angular/29808/workflows/32ea144f-3273-4de2-9318-f1d08a605cae/jobs/941619/artifacts). In particular, you'll need `@angular/compiler` and `@angular/compiler-cli`. These are based on 12.0 preview builds and only TS 4.2 is supported so you could either update TS to 4.2.x _or_ set `disableTypeScriptVersionCheck: true` in `angularCompilerOptions` in a `tsconfig.json` file. If you encounter too much trouble with this I can probably cherry-pick it to 11.2.x, please let me know if that's desired.

Using those build artifacts you should get slightly better AOT times (I estimate ~400ms less). Next, I'm interested in a more extensive optimization of which I have no idea how effective it is. This is hidden behind the `NG_EXPERIMENTAL_CYCLES` environment variable; if you set that to `1` it will take a different approach for cycle detection. I'd also like how that compares. If you can, please share AOT / NO CIRC / NO SM profiles using those build artifacts with and without `NG_EXPERIMENTAL_CYCLES=1`.

Instructions for installing the build artifacts [can be found here](https://github.com/angular/angular/blob/master/docs/DEVELOPER.md#archives-for-each-package).

---

Unrelated to the above, the CLI is ramping up for the adoption of Webpack 5 in v12. Your project is a gem in terms of the insights it provides, so if you could have a go at Webpack 5 with the CLI's latest `next` releases that would also be super valuable. I think that using Webpack 5 does require a Yarn `resolutions` rule, though, so I'm not sure if testing that is feasible for you.

**Comment by montella1507** [comment]
Will do at weekend.

**Comment by JoostK** [comment]
Oh and one more thing that we noticed in the profile that deserves further research: in AOT mode the Angular compiler does template type-checking by feeding the Angular templates into TypeScript. This is done by injecting additional source files into a temporary TypeScript program. Creating that temporary program should be efficient as it's mostly identical to the original. In your profile however, we're seeing a codepath that indicates that TypeScript has determined that it can't reuse everything from the original program, which incurs an extra cost (circa 500ms as taken from your profiles.). So to verify: are the edits you're making entirely redundant, such as whitespace only changes? They probably are, in which case we'd need to do some experimenting to figure out if we can avoid that extra cost (we have tests that tell us this should work, but perhaps it's missing a case).

**Comment by DavidDudson** [comment]
Hi everyone! Thank you all for sorting these issues out, our team is currently switching from 9 => 11 and are holding off for these performance improvements at present. Do we know roughly what compile times are, comparable to Angular 9/10, rather than just earlier Angular 11 versions? I'll be doing some experimentations next week profiling our codebase with the lastest angular version.

**Comment by montella1507** [comment]
> Oh and one more thing that we noticed in the profile that deserves further research: in AOT mode the Angular compiler does template type-checking by feeding the Angular templates into TypeScript. This is done by injecting additional source files into a temporary TypeScript program. Creating that temporary program should be efficient as it's mostly identical to the original. In your profile however, we're seeing a codepath that indicates that TypeScript has determined that it can't reuse everything from the original program, which incurs an extra cost (circa 500ms as taken from your profiles.). So to verify: are the edits you're making entirely redundant, such as whitespace only changes? They probably are, in which case we'd need to do some experimenting to figure out if we can avoid that extra cost (we have tests that tell us this should work, but perhaps it's missing a case).

i do everytime the same - editing string / number value. eg.   return "george";  >> return "georgedcjwekwejkfk"; etc. :-) I can do bigger changes, if you wish.

**Comment by montella1507** [comment]
> Hi everyone! Thank you all for sorting these issues out, our team is currently switching from 9 => 11 and are holding off for these performance improvements at present. Do we know roughly what compile times are, comparable to Angular 9/10, rather than just earlier Angular 11 versions? I'll be doing some experimentations next week profiling our codebase with the lastest angular version.

as far as i know... Angular rebuild times went worse and worse from 8 -> 11 in all projects i participate to...  HOWEVER... latest version is deal breaker + we have realised how BAD is circular detection in our bigger projects (rebuild times dropped by 80% by disabling that check)

**Comment by JoostK** [comment]
Nope, the current profiles are sufficient. My question was to ensure that you're indeed only making trivial changes. TypeScript has some logic in `tryReuseStructureFromOldProgram` to determine to what extend the structure of a previous program can be reused. That should be `Completely` for the template type-check program, but your profiles show it's not. I can't tell whether it's `Not` or `SafeModules` (the other variants of structure reuse). It would be super helpful to know which code-paths are hit there that prevent complete structure reuse, but getting that information out requires some debugging and that's not trivial for me to guide you through. The simplest way would be to hack in a lot of `console.log`'s into all places where `structureIsReused` is set or the function early-returns, or to attach a debugger and set breakpoints at these places to find out where it's hit. The `console.log` approach is probably much simpler to share the results with, though...

**Comment by montella1507** [comment]
Look, we have 2 options...   

1) I can put console.log anywhere you want
2) I can provide Remote desktop for you and you can do it yourself (public IP / powerfull PC / powerfull connection). You are in the same timezone i guess (Netherland / UTC+1) so it should not be a problem.. We can do that this weekend or other day if you willing to.

**Comment by JoostK** [comment]
Let's keep it offline :-) I am on [Angular's Discord](discord.gg/angular)  as `Joost#9883` if you want to chat.

**Comment by montella1507** [comment]
Discussed  on discord. Will let this issue open, @JoostK  close when you think all is solved :D

**Comment by montella1507** [comment]
TBH i think i should have closed this issue, because it was resolved.

**Comment by gabrielguerrero** [comment]
@montella1507 sorry I think is fine to close it, nx just released a new release , and it seems to be working fine with it , build is back again in 1 min using latest angular-build, I think they have a custom builder that wraps yours and not sure exactly what went wrong there, in their change log looks like they fixed a bug were ngcc was been called on incremental builds, I though it was related to this because like I said the issue seem to be related to angular-build in my test. Anyway sorry to bother you, I spend so many hours investigating this , thanks anyway.

**Comment by JoostK** [comment]
In the meantime we have also addressed another slow path that became clear from @montella1507's profile, albeit only for v12. Closing this issue as not further actionable.
